[SPARK-28344][SQL] detect ambiguous self-join and fail the query

cloud-fan opened the pull request:
## What changes were proposed in this pull request?

This is an alternative solution of https://github.com/apache/spark/pull/24442 . It fails the query if ambiguous self join is detected, instead of trying to disambiguate it. The problem is that, it's hard to come up with a reasonable rule to disambiguate, the rule proposed by #24442 is mostly a heuristic.

### background of the self-join problem:
This is a long-standing bug and I've seen many people complaining about it in JIRA/dev list.

A typical example:
```
val df1 = …
val df2 = df1.filter(...)
df1.join(df2, df1("a") > df2("a")) // returns empty result
```
The root cause is, `Dataset.apply` is so powerful that users think it returns a column reference which can point to the column of the Dataset at anywhere. This is not true in many cases. `Dataset.apply` returns an `AttributeReference` . Different Datasets may share the same `AttributeReference`. In the example above, `df2` adds a Filter operator above the logical plan of `df1`, and the Filter operator reserves the output `AttributeReference` of its child. This means, `df1("a")` is exactly the same as `df2("a")`, and `df1("a") > df2("a")` always evaluates to false.

### The rule to detect ambiguous column reference caused by self join:
We can reuse the infra in #24442 :
1. each Dataset has a globally unique id.
2. the `AttributeReference` returned by `Dataset.apply` carries the ID and column position(e.g. 3rd column of the Dataset) via metadata.
3. the logical plan of a `Dataset` carries the ID via `TreeNodeTag`

When self-join happens, the analyzer asks the right side plan of join to re-generate output attributes with new exprIds. Based on it, a simple rule to detect ambiguous self join is:
1. find all column references (i.e. `AttributeReference`s with Dataset ID and col position) in the root node of a query plan.
2. for each column reference, traverse the query plan tree, find a sub-plan that carries Dataset ID and the ID is the same as the one in the column reference.
3. get the corresponding output attribute of the sub-plan by the col position in the column reference.
4. if the corresponding output attribute has a different exprID than the column reference, then it means this sub-plan is on the right side of a self-join and has regenerated its output attributes. This is an ambiguous self join because the column reference points to a table being self-joined.


## How was this patch tested?

existing tests and new test cases

cloud-fan committed:
detect ambiguous self-join and fail the query

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:
```diff
@@ -66,30 +66,24 @@ class QueryExecution(
     assertAnalyzed()
     assertSupported()
-    sparkSession.sharedState.cacheManager.useCachedData(analyzed)
+    sparkSession.sharedState.cacheManager.useCachedData(analyzed.clone())
```
I'll move this out as a separate PR.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12631/
Test PASSed.

SparkQA commented:
**[Test build #107505 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107505/testReport)** for PR 25107 at commit [`1898674`](https://github.com/apache/spark/commit/189867428b8851da5242765641c2967608f1a833).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107505/
Test FAILed.

cloud-fan commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12650/
Test PASSed.

SparkQA commented:
**[Test build #107524 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107524/testReport)** for PR 25107 at commit [`1898674`](https://github.com/apache/spark/commit/189867428b8851da5242765641c2967608f1a833).

SparkQA commented:
**[Test build #107524 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107524/testReport)** for PR 25107 at commit [`1898674`](https://github.com/apache/spark/commit/189867428b8851da5242765641c2967608f1a833).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DetectAmbiguousSelfJoin(conf: SQLConf) extends Rule[LogicalPlan] `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107524/
Test FAILed.

SparkQA commented:
**[Test build #107535 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107535/testReport)** for PR 25107 at commit [`69627b4`](https://github.com/apache/spark/commit/69627b4aed7a02aad4ca5a96136f907214dc3e35).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12661/
Test PASSed.

SparkQA commented:
**[Test build #107535 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107535/testReport)** for PR 25107 at commit [`69627b4`](https://github.com/apache/spark/commit/69627b4aed7a02aad4ca5a96136f907214dc3e35).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DetectAmbiguousSelfJoin(conf: SQLConf) extends Rule[LogicalPlan] `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107535/
Test PASSed.

albertusk95 commented on sql/core/src/main/scala/org/apache/spark/sql/execution/analysis/DetectAmbiguousSelfJoin.scala:
```diff
@@ -0,0 +1,159 @@
+          "to figure out which one. Please alias the Datasets with different names via " +
+          "`Dataset.as` before joining them, and specify the column using qualified name, e.g. " +
+          """`df.as("a").join(df.as("b"), $"a.id" > $"b.id")`. You can also set """ +
+          s"${SQLConf.FAIL_AMBIGUOUS_SELF_JOIN} to false to disable this check.")
```
it seems that based on my experience, aliasing the dataset before joining still results in an `ambiguous reference exception` when a certain column is selected. For instance, `joined_df = df.as("a").join(df.as("b"), $"a.id" > $"b.id")` and then `joined_df.select('certain_column')` gave an exception.

Providing the alias name didn't help as well -> `joined_df.select(a.certain_column)`.

However, by deep copying the dataframes gave the correct result.

cloud-fan replied:
Can you give a concrete example? It looks to me that `joined_df.select($"a.certain_column")` should work.

albertusk95 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -787,6 +787,13 @@ object SQLConf {
+      .booleanConf
+      .createWithDefault(true)
+
   // Whether to retain group by columns or not in GroupedData.agg.
```
I think moving this comment to `.doc()` like the above is better

cloud-fan replied:
This is unrelated to this patch, we can fix it and other similar configs in another PR.

SparkQA commented:
**[Test build #108076 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108076/testReport)** for PR 25107 at commit [`6bff6f3`](https://github.com/apache/spark/commit/6bff6f32a6a1ccd237e0f33ae9228562fb8932b2).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13182/
Test PASSed.

SparkQA commented:
**[Test build #108076 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108076/testReport)** for PR 25107 at commit [`6bff6f3`](https://github.com/apache/spark/commit/6bff6f32a6a1ccd237e0f33ae9228562fb8932b2).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DetectAmbiguousSelfJoin(conf: SQLConf) extends Rule[LogicalPlan] `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108076/
Test FAILed.

cloud-fan commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13201/
Test PASSed.

SparkQA commented:
**[Test build #108093 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108093/testReport)** for PR 25107 at commit [`6bff6f3`](https://github.com/apache/spark/commit/6bff6f32a6a1ccd237e0f33ae9228562fb8932b2).

SparkQA commented:
**[Test build #108093 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108093/testReport)** for PR 25107 at commit [`6bff6f3`](https://github.com/apache/spark/commit/6bff6f32a6a1ccd237e0f33ae9228562fb8932b2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DetectAmbiguousSelfJoin(conf: SQLConf) extends Rule[LogicalPlan] `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108093/
Test PASSed.

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -1312,10 +1326,24 @@ class Dataset[T] private[sql](
       } else {
         val expr = resolve(colName)
-        Column(expr)
+        Column(addDataFrameIdToCol(expr))
```
This is the only place that requires calling `addDataFrameIdToCol`. How about colRegex? 

We might need a clear comment and explain which cases we should call it?

maryannxue replied:
Same question: doesn't it apply to `colRegex`?

cloud-fan replied:
good catch! We should apply it in `colRegex` as well.

maryannxue commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -794,6 +794,13 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val FAIL_AMBIGUOUS_SELF_JOIN =
```
Would `FAIL_AMBIGUOUS_DATASET_COLUMN_REF` be more accurate?
And in the doc below, we can describe when this usually happens (e.g., with self-join).

cloud-fan replied:
A self-join that contains ambiguous column reference is an ambiguous self-join. I think ambiguous self-join is shorter and easier to understand, as a config name.

cloud-fan replied:
BTW there are no other places that can make column reference ambiguous. Self-join is the only place.

maryannxue commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameJoinSuite.scala:
```diff
@@ -164,6 +175,44 @@ class DataFrameJoinSuite extends QueryTest with SharedSQLContext {
       Row(1, 1, 1, 1) :: Row(2, 1, 2, 2) :: Nil)
   }
 
+  test("SPARK-28344: fail ambiguous self join") {
```
Let's do a dedicated suite have individual cases for these scenarios.

maryannxue commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameJoinSuite.scala:
```diff
@@ -164,6 +175,44 @@ class DataFrameJoinSuite extends QueryTest with SharedSQLContext {
+      assertAmbiguousSelfJoin(df3.join(df1.join(df2), df2("id") < df3("id")))
+    }
+  }
+
```
Can we add more tests to cover:
1. aliased references work fine;
2. suppose df4 has a different attribute set, and test `df1.join(df4).df2.select(df1("id"))`?

dilipbiswal commented on docs/sql-migration-guide-upgrade.md:
```diff
@@ -155,12 +155,14 @@ license: |
 
   - Since Spark 3.0, 0-argument Java UDF is executed in the executor side identically with other UDFs. In Spark version 2.4 and earlier, 0-argument Java UDF alone was executed in the driver side, and the result was propagated to executors, which might be more performant in some cases but caused inconsistency with a correctness issue in some cases.
 
+  - Since Spark 3.0, Dataset query fails if it contains ambiguous column reference that is caused by self join. A typical example: `val df1 = ...; val df2 = df1.filter(...);`, then `df1.join(df2, df1("a") > df2("a"))` returns empty result which is quite confusing. This is because Spark cannot resolve Dataset column references that point to tables being self joined, and `df1("a")` is exactly the s ...
```
@cloud-fan Should we take the opportunity to describe the other way to correct the problem is by using column aliases (with an example) ? Or do you think that would make the text too long for users to consume ?

cloud-fan replied:
> describe the other way to correct the problem is by using column aliases (with an example)

The error message contains it, see https://github.com/apache/spark/pull/25107/files#diff-72682666ae0e00b0be514f6867838be5R143

dilipbiswal replied:
@cloud-fan Ah.. got it.. Thanks Wenchen.

cloud-fan committed:
address comments

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13402/
Test PASSed.

SparkQA commented:
**[Test build #108305 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108305/testReport)** for PR 25107 at commit [`4870abc`](https://github.com/apache/spark/commit/4870abc4cac49672fa164eccdd7459a05a9ac969).

SparkQA commented:
**[Test build #108305 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108305/testReport)** for PR 25107 at commit [`4870abc`](https://github.com/apache/spark/commit/4870abc4cac49672fa164eccdd7459a05a9ac969).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108305/
Test PASSed.

gatorsmile commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameSelfJoinSuite.scala:
```diff
@@ -0,0 +1,183 @@
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.test.SharedSparkSession
+
+class DataFrameSelfJoinSuite extends QueryTest with SharedSparkSession {
```
Also need a test case for colRegex

cloud-fan committed:
add one more test

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13473/
Test PASSed.

SparkQA commented:
**[Test build #108378 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108378/testReport)** for PR 25107 at commit [`65b6762`](https://github.com/apache/spark/commit/65b6762423822b79ac2fc21d80a02679701ebf29).

SparkQA commented:
**[Test build #108378 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108378/testReport)** for PR 25107 at commit [`65b6762`](https://github.com/apache/spark/commit/65b6762423822b79ac2fc21d80a02679701ebf29).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108378/
Test PASSed.

gatorsmile commented:
How about the nested column? 
```Scala
    val df = spark.read
      .json(Seq("""{"a": {"b": 1, "a": {"a": 1}}, "c": [{"d": 1}]}""").toDS())
    val df1 = df
    df1.join(df, df("a.b") > df1("a.b"))
```

cloud-fan commented:
The ambiguity comes from `AttributeReference`s with the same expr ID. The nested column in Spark is just an expression: `ExtractValue(Attr, ...)`. It's not related to ambiguous self join, like other expressions.

Your example is the same as
```
val df1 = ...
val df2 = df1.filter...
df1.join(df2, df1("a") * 3 > df2("a") + 2)
```
which is definitely a ambiguous self-join because the column `a` is ambiguous in the query.

cloud-fan committed:
fix

cloud-fan commented:
`df("a.b")` is different from things like `df("a") + 1`, because we need to take care of `df("a.b")` to attach the dataframe id. I've fixed it and added a test. @gatorsmile thanks for catching it!

SparkQA commented:
**[Test build #108474 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108474/testReport)** for PR 25107 at commit [`6b18efe`](https://github.com/apache/spark/commit/6b18efe29199c5b9bb574a4a3c50a6e3a89f557a).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13568/
Test PASSed.

SparkQA commented:
**[Test build #108474 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108474/testReport)** for PR 25107 at commit [`6b18efe`](https://github.com/apache/spark/commit/6b18efe29199c5b9bb574a4a3c50a6e3a89f557a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108474/
Test PASSed.

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/Column.scala:
```diff
@@ -144,11 +153,16 @@ class Column(val expr: Expression) extends Logging {
-  override def hashCode: Int = this.expr.hashCode()
+  override def hashCode: Int = this.normalizedExpr().hashCode()
+
+  private def normalizedExpr(): Expression = expr match {
```
When we add the metadata in addDataFrameIdToCol, we use transform. Do we need to use it here when removing it? 

Our test case can cover it?

cloud-fan replied:
good catch!

However, I'm not able to test it. The `normalizedExpr` is only used in `equals` and `hashCode`. For `df("a.b")`, we will create a new alias everytime, see `AttributeSeq.resolve`.

That said, if we get multiple `Column` instances via calling `df("a.b")` multiple times, these `Column` instances will never equal to each other, because the alias is different.

cloud-fan committed:
fix

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13610/
Test PASSed.

SparkQA commented:
**[Test build #108516 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108516/testReport)** for PR 25107 at commit [`eec92f6`](https://github.com/apache/spark/commit/eec92f68e9c80ee88b4c7ef6613d2e5a31c80138).

SparkQA commented:
**[Test build #108516 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108516/testReport)** for PR 25107 at commit [`eec92f6`](https://github.com/apache/spark/commit/eec92f68e9c80ee88b4c7ef6613d2e5a31c80138).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108516/
Test PASSed.

gatorsmile commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionStateBuilder.scala:
```diff
@@ -78,7 +79,8 @@ class HiveSessionStateBuilder(session: SparkSession, parentState: Option[Session
 
     override val postHocResolutionRules: Seq[Rule[LogicalPlan]] =
-      new DetermineTableStats(session) +:
+      new DetectAmbiguousSelfJoin(conf) +:
```
How about adding a test case for ensuring all these rules are included? So far, we do not have a test case for covering it if the rule is missing to add.

cloud-fan replied:
I need to know the details of each rule so that I can write a test to make sure these rules are included. Maybe do it in a followup instead of blocking this PR? BTW the new tests added in this PR do guarantee the `DetectAmbiguousSelfJoin` rule is there.

gatorsmile replied:
Created a ticket https://issues.apache.org/jira/browse/SPARK-28629

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/analysis/DetectAmbiguousSelfJoin.scala:
```diff
@@ -0,0 +1,159 @@
+ * [[AttributeReference]] with new expr IDs for the right side plan of the join. If the Dataset
+ * column reference points to a column in the right side plan of a self-join, users will get
+ * unexpected result because the column reference can't match the newly generated
+ * [[AttributeReference]].
```
In the description, we need to also document the metadata are removed by this rule.

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -1311,11 +1325,27 @@ class Dataset[T] private[sql](
 
+  // Attach the dataset id and column position to the column reference, so that we can detect
+  // ambiguous self-join correctly.See the rule `DetectAmbbiguousSelfJoin`.
+  // This must be called before we return a `Column` that contains `AttributeReference`.
```
Also, document these metadata will be removed in the rule DetectAmbbiguousSelfJoin

gatorsmile commented:
LGTM except the above comments. cc @maryannxue

cloud-fan committed:
Merge remote-tracking branch 'origin/master' into new-self-join

cloud-fan committed:
address comments

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13732/
Test PASSed.

SparkQA commented:
**[Test build #108645 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108645/testReport)** for PR 25107 at commit [`62228e8`](https://github.com/apache/spark/commit/62228e8ffbb5d54ad2ec5a85dd2450cd1dec5ae2).

SparkQA commented:
**[Test build #108645 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108645/testReport)** for PR 25107 at commit [`62228e8`](https://github.com/apache/spark/commit/62228e8ffbb5d54ad2ec5a85dd2450cd1dec5ae2).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108645/
Test FAILed.

cloud-fan commented:
retest this please

SparkQA commented:
**[Test build #108657 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108657/testReport)** for PR 25107 at commit [`62228e8`](https://github.com/apache/spark/commit/62228e8ffbb5d54ad2ec5a85dd2450cd1dec5ae2).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13743/
Test PASSed.

SparkQA commented:
**[Test build #108657 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108657/testReport)** for PR 25107 at commit [`62228e8`](https://github.com/apache/spark/commit/62228e8ffbb5d54ad2ec5a85dd2450cd1dec5ae2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108657/
Test PASSed.

gatorsmile commented:
LGTM

cloud-fan commented:
thanks, merging to master!

nchammas commented:
I wonder if this patch fixes the correctness issue reported (with repro) in [SPARK-25150](https://issues.apache.org/jira/browse/SPARK-25150). I'll give it a try later this week.

dongjoon-hyun commented:
Hi, All.
Can we have this in branch-2.4?
(Also, cc @tgravescs )

cloud-fan commented:
This is not a complete fix but fail the query if we are not able to resolve the ambiguity. Is it OK to backport? It can break existing queries.

dongjoon-hyun commented:
I think we are safe because we have `spark.sql.analyzer.failAmbiguousSelfJoin`.

Do we have a false positive?
1. If this PR makes a correct existing query fail, we can backport this patch with `spark.sql.analyzer.failAmbiguousSelfJoin=false` by default.
2. If this PR prevents only wrong query, we can backport this patch like 3.0. `spark.sql.analyzer.failAmbiguousSelfJoin=true` by default.

In both (1) and (2) case, we can claim that the correctness issue is resolved.
How do you think about that? (Also, cc @gatorsmile )

cloud-fan commented:
I'm fine to backport. I'm on vacation soon, can someone help to backport?

dongjoon-hyun commented:
Hi, @cloud-fan . Did you come back from the vacation? :)

cloud-fan commented:
done in https://github.com/apache/spark/pull/27417

dongjoon-hyun commented:
It turns out there are at least 5 commits required. Please see the discussion on #27417 . We are re-evaluating the risks.
